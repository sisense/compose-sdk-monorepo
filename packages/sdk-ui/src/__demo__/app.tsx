import { Alert, Tab, Tabs } from '@mui/material';
import { Box } from '@mui/system';
import { useState } from 'react';
import { SisenseContextProvider } from '../components/SisenseContextProvider';
import { SisenseContextProviderProps } from '../props';
import { ChartsFromExampleApp } from './pages/ChartsFromExampleApp';
import { ECommerceDemo } from './pages/ECommerceDemo';
import { MiscDemo } from './pages/MiscDemo';
import { WidgetDemo } from './pages/WidgetDemo';

// This page is meant to enable faster iterations during development than
// using react-ts-demo or other demo apps that require a built sdk-ui
// Simply add a page to the `pages` array and it will show up in `yarn dev`
// Suggest adding a router or at least sessionStorage var for selectedTabIndex
// if this becomes popular
const pages = [WidgetDemo, ECommerceDemo, ChartsFromExampleApp, MiscDemo];

const { VITE_APP_SISENSE_URL, VITE_APP_SISENSE_USERNAME, VITE_APP_SISENSE_PASSWORD } = import.meta
  .env;

export function App() {
  const [selectedTabIndex, setSelectedTabIndex] = useState(0);

  const shouldShowAlert =
    !VITE_APP_SISENSE_URL || !VITE_APP_SISENSE_USERNAME || !VITE_APP_SISENSE_PASSWORD;

  const sisenseContextProps: SisenseContextProviderProps = {
    url: VITE_APP_SISENSE_URL ?? '',
    username: VITE_APP_SISENSE_USERNAME,
    password: VITE_APP_SISENSE_PASSWORD,
    defaultDataSource: 'Sample ECommerce',
  };

  return (
    <>
      {shouldShowAlert && (
        <Alert severity="warning">
          One or more expected env vars are missing. You may want to check your
          <code>.env.local</code> file.
        </Alert>
      )}
      <SisenseContextProvider {...sisenseContextProps}>
        <div style={{ width: '100vw', height: '100vh' }}>
          <Box sx={{ borderBottom: 1, borderColor: 'divider' }}>
            <Tabs
              value={selectedTabIndex}
              onChange={(e, value: number) => setSelectedTabIndex(value)}
            >
              {pages.map((page, i) => (
                <Tab key={i} label={page.name} sx={{ textTransform: 'none' }} />
              ))}
            </Tabs>
          </Box>
          {pages.map((Page, i) => (
            <div key={i} hidden={selectedTabIndex !== i}>
              {selectedTabIndex === i && (
                <Box sx={{ p: 3 }}>
                  <Page />
                </Box>
              )}
            </div>
          ))}
        </div>
      </SisenseContextProvider>
    </>
  );
}
