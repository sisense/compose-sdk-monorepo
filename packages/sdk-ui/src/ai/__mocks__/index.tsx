import { SisenseContextProvider } from '@/sisense-context/sisense-context-provider';
import { mockToken, mockUrl } from '@/__mocks__/msw';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { ReactNode } from 'react';
import AiContextProvider from '../ai-context-provider';

/**
 * A component that wraps an AI hook/component with all required providers.
 * For use in unit tests only.
 *
 * @internal
 */
export const AiTestWrapper = ({ children }: { children: ReactNode }) => (
  <SisenseContextProvider url={mockUrl} token={mockToken}>
    <AiContextProvider>
      <QueryClientProvider
        client={
          new QueryClient({
            defaultOptions: {
              queries: {
                retry: false,
              },
            },
          })
        }
      >
        {children}
      </QueryClientProvider>
    </AiContextProvider>
  </SisenseContextProvider>
);
