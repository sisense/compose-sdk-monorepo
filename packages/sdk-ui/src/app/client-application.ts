/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-argument */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable max-params */
import { HttpClient, getAuthenticator, isWatAuthenticator } from '@sisense/sdk-rest-client';
import { DimensionalQueryClient, QueryClient } from '@sisense/sdk-query-client';
import { DataSource } from '@sisense/sdk-data';
import { PivotClient } from '@sisense/sdk-pivot-client';
import { TrackingEventDetails } from '@sisense/sdk-tracking';
import { normalizeUrl } from '@sisense/sdk-common';
import { DateConfig } from '../query/date-formats';
import { AppSettings, getSettings } from './settings/settings';
import { TranslatableError } from '../translation/translatable-error';
import { TranslationConfig, LoadingIndicatorConfig } from '../types';
import { clearExecuteQueryCache } from '@/query/execute-query';
import { SisenseContextProviderProps } from '@/props';
import { SYSTEM_TENANT_NAME } from '@/const';

/**
 * Application configuration
 */
export type AppConfig = {
  /**
   * A [date-fns Locale](https://date-fns.org/v2.30.0/docs/Locale)
   */
  locale?: Locale;

  /**
   * Translation Configuration
   */
  translationConfig?: TranslationConfig;

  /**
   * Date Configurations
   */
  dateConfig?: DateConfig;

  /**
   * Loading Indicator Configurations
   */
  loadingIndicatorConfig?: LoadingIndicatorConfig;

  /**
   * Query Cache Configurations.
   *
   * See [Client query caching guide](/guides/sdk/guides/client-query-caching.html) for more details.
   *
   * @alpha
   */
  queryCacheConfig?: {
    /**
     * Whether to enable client-side query caching.
     *
     * If not specified, the default value is `false`
     */
    enabled?: boolean;
  };

  /**
   * Query limit (max rows count that will be fetched in query)
   *
   * @default 20000
   */
  queryLimit?: number;

  /**
   * Accessibility configuration. Set the `accessibilityConfig.enabled` property to `true` to enable accessibility features for charts built with Highcharts.
   *
   * Once the accessibility configuration in enabled, you can use the default descriptions or choose to create custom configurations for the descriptions of a chart, its axes, its series, and values description formatting. To create custom configurations, use the `highchartsOptions` object that is passed to the {@link BeforeRenderHandler} of the chart's `onBeforeRender` callback. Modify the object using the accessibility options as described in the [Accessibility module documentation](https://www.highcharts.com/docs/accessibility/accessibility-module).
   *
   * Note that enabling accessibility also causes markers to appear in charts even if they are disabled using the chart's `styleOptions`.
   *
   * This feature is in alpha.
   */
  accessibilityConfig?: {
    /**
     * Whether to enable accessibility features
     *
     * If not specified, the default value is `false`
     */
    enabled?: boolean;
  };

  /**
   * Configuration of the tabber widget
   *
   * Set the `tabberConfig.enabled` property to `true` to enable the tabber widget support. Otherwise, the tabber widget will be treated as 'unknown custom widget type'.
   * Only tabber widgets that are configured using the Fusion UI as documented [here](https://docs.sisense.com/main/SisenseLinux/tabber.htm#Creating) are supported.
   */
  tabberConfig?: {
    /**
     * Whether to enable tabber widget support
     *
     * If not specified, the default value is `true`
     */
    enabled?: boolean;
  };
  /**
   * Configuration of the component that is rendered in case of an error
   */
  errorBoundaryConfig?: {
    /**
     * Whether to show error text without hovering over the error icon
     *
     * If not specified, the default value is `false`
     */
    alwaysShowErrorText?: boolean;
  };

  /**
   * Tracking configuration
   */
  trackingConfig?: {
    /**
     * Whether to enable or disable tracking in development or test environment.
     *
     * If not specified, the default value is `true`
     *
     * In production, tracking is always enabled.
     *
     * @internal
     */
    enabled?: boolean;
    /**
     * Callback to be invoked when tracking event occurs
     */
    onTrackingEvent?: (payload: TrackingEventDetails) => void;
  };

  /**
   * Configuration of the jump to dashboard feature
   * Only the widgets with pre-configured JTD config will support this feature.
   * You can configure the JTD config in the widget settings in Fusion UI having jump to dashboard plugin enabled.
   *
   * This feature is currently in alpha.
   * To learn more about the jump to dashboard feature, see the [Jump to Dashboard documentation](https://docs.sisense.com/main/SisenseLinux/jump-to-dashboard.htm?tocpath=Add-ons%7C_____2#ConfiguringJTDPerWidget).
   */
  jumpToDashboardConfig?: {
    /**
     * Whether to enable or disable the jump to dashboard feature
     *
     * If not specified, the default value is `false`
     */
    enabled?: boolean;
  };
};

/**
 * Stands for a Sisense Client Application which connects to a Sisense Environment
 *
 * @internal
 */
export interface ClientApplication {
  /**
   * Gets the underlying HTTP Client
   */
  readonly httpClient: HttpClient;

  /**
   * Gets the underlying Pivot Client
   */
  readonly pivotClient: PivotClient;

  /**
   * Gets the underlying Query Client
   */
  readonly queryClient: QueryClient;

  /**
   * Gets the default data source being used as default for child components with no explicitly defined data source
   */
  readonly defaultDataSource?: DataSource;

  /**
   * Gets the application settings
   */
  settings: AppSettings;

  /**
   * Gets the module to control query cache
   */
  queryCache: {
    /**
     * Clears the query cache
     */
    clear: () => void;
  };
}

type ClientApplicationParams = Pick<
  SisenseContextProviderProps,
  | 'appConfig'
  | 'defaultDataSource'
  | 'url'
  | 'token'
  | 'wat'
  | 'ssoEnabled'
  | 'enableSilentPreAuth'
  | 'useFusionAuth'
>;

function getBaseUrl(url: string, tenantName: string) {
  return tenantName === SYSTEM_TENANT_NAME ? url : url.replace(`/${tenantName}`, '');
}

function normalizeErrors() {
  // todo: add unsubscribe flow
  window.addEventListener('unhandledrejection', (event: PromiseRejectionEvent) => {
    // TODO: replace this with custom logger
    if (event.reason instanceof Error) {
      console.error(event.reason.message);
    } else {
      console.error(event.reason);
    }
  });
}

/** @internal */
export const createClientApplication = async ({
  defaultDataSource,
  url: rawUrl,
  token,
  wat,
  ssoEnabled,
  appConfig,
  enableSilentPreAuth,
  useFusionAuth,
}: ClientApplicationParams): Promise<ClientApplication> => {
  if (rawUrl === undefined) {
    throw new TranslatableError('errors.sisenseContextNoAuthentication');
  }

  const url = normalizeUrl(rawUrl);
  const urlWithSearchParams = normalizeUrl(rawUrl, true);

  const auth = getAuthenticator({
    url: urlWithSearchParams,
    token,
    wat,
    ssoEnabled,
    enableSilentPreAuth,
    useFusionAuth,
  });

  if (!auth) {
    throw new TranslatableError('errors.sisenseContextNoAuthentication');
  }

  normalizeErrors();

  const httpClient = new HttpClient(
    url,
    auth,
    'sdk-ui' + (__PACKAGE_VERSION__ ? `-${__PACKAGE_VERSION__}` : ''),
  );
  const loginSuccess = await httpClient.login();

  // do not fetch palette settings from server if login failed
  // SSO redirect is considered failed login as there will be another login attempt
  // TODO: Remove WAT check once the server will be able to return the palette under the WAT
  const useDefaultPalette = isWatAuthenticator(auth) || !loginSuccess;
  const settings = await getSettings(appConfig || {}, httpClient, useDefaultPalette);

  const pivotClient = new PivotClient(getBaseUrl(url, settings.user.tenant.name), auth);
  const queryClient = new DimensionalQueryClient(httpClient, pivotClient);

  const queryCache = {
    clear: clearExecuteQueryCache,
  };

  return {
    httpClient,
    pivotClient,
    queryClient,
    settings,
    // todo: make it optional (incorrect previous implementation)
    defaultDataSource: defaultDataSource,
    queryCache,
  };
};
