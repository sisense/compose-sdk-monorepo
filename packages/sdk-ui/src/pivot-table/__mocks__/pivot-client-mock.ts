/* eslint-disable no-unused-vars */
/* eslint-disable sonarjs/no-identical-functions */
import React from 'react';
import { PivotClient } from '@ethings-os/sdk-pivot-client';

// Mock of PivotBuilder class
class MockPivotBuilder {
  private listeners: Record<string, Array<(...args: any[]) => void>> = {};

  on(eventName: string, callback: (...args: any[]) => void) {
    if (!this.listeners[eventName]) {
      this.listeners[eventName] = [];
    }
    this.listeners[eventName].push(callback);
  }

  off(eventName: string, callback: (...args: any[]) => void) {
    if (this.listeners[eventName]) {
      this.listeners[eventName] = this.listeners[eventName].filter((cb) => cb !== callback);
    }
  }

  emit(eventName: string, ...args: any[]) {
    if (this.listeners[eventName]) {
      this.listeners[eventName].forEach((callback) => callback(...args));
    }
  }

  render(props: any) {
    // Mock render method - does not actually render anything
    // Simulate emitting a proper React element instead of a string
    const mockElement = React.createElement(
      'div',
      { 'data-testid': 'mocked-pivot-table' },
      'Mocked Pivot Table',
    );
    setTimeout(() => {
      this.emit('pivotElementChange', mockElement);
    }, 0);
  }

  updateJaql(jaql: any) {
    // Mock updateJaql method
    setTimeout(() => {
      this.emit('queryStart');
      // Simulate query end with empty metadata
      setTimeout(() => {
        this.emit('queryEnd', { cellsMetadata: null });
      }, 10);
    }, 0);
  }

  updateDataService(dataService: any) {
    // Mock updateDataService method
  }

  destroy() {
    // Mock destroy method
    this.listeners = {};
  }
}

// Mock of DataService class
class MockDataService {
  private listeners: Record<string, Array<(...args: any[]) => void>> = {};

  on(eventName: string, callback: (...args: any[]) => void) {
    if (!this.listeners[eventName]) {
      this.listeners[eventName] = [];
    }
    this.listeners[eventName].push(callback);
  }

  off(eventName: string, callback: (...args: any[]) => void) {
    if (this.listeners[eventName]) {
      this.listeners[eventName] = this.listeners[eventName].filter((cb) => cb !== callback);
    }
  }

  destroy() {
    // Mock destroy method
    this.listeners = {};
  }
}

// Create a mocked PivotClient
export const createMockPivotClient = () => {
  return {
    preparePivotBuilder: vi.fn(() => new MockPivotBuilder()),
    prepareDataService: vi.fn(() => new MockDataService()),
    queryData: vi.fn(),
    socketBuilder: {
      socket: {},
    },
  } as unknown as PivotClient;
};
