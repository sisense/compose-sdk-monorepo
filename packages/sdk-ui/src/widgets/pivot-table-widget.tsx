import { useState, type FunctionComponent } from 'react';
import { PivotTableWidgetProps } from '../props';
import { asSisenseComponent } from '../decorators/component-decorators/as-sisense-component';
import { DynamicSizeContainer, getWidgetDefaultSize } from '../dynamic-size-container';
import { getDataSourceName } from '@sisense/sdk-data';
import { WidgetContainer } from './common/widget-container';
import { useSisenseContext } from '@/sisense-context/sisense-context';
import { PivotTable } from '../pivot-table';
import { DEFAULT_WIDGET_HEADER_HEIGHT } from './constants';

function calcPivotTableWidgetHeight(pivotTableHeight: number | undefined) {
  return pivotTableHeight ? pivotTableHeight + DEFAULT_WIDGET_HEADER_HEIGHT : undefined;
}

/**
 * React component extending `PivotTable` to support widget style options.
 *
 * @example
 * Example of using the `PivotTableWidget` component to
 * plot a pivot table over the `Sample ECommerce` data source hosted in a Sisense instance.
 * ```tsx
 * <PivotTableWidget
 *   dataSource={DM.DataSource}
 *   dataOptions={{
 *     rows: [DM.Category.Category],
 *     values: [measureFactory.sum(DM.Commerce.Cost, 'Total Cost')]
 *   }}
 * />
 * ```
 * @param props - Pivot Table Widget properties
 * @returns Widget component representing a pivot table
 * @group Dashboards
 * @beta
 */
export const PivotTableWidget: FunctionComponent<PivotTableWidgetProps> = asSisenseComponent({
  componentName: 'PivotTableWidget',
})((props) => {
  const [refreshCounter, setRefreshCounter] = useState(0);
  const [pivotTableHeight, setPivotTableHeight] = useState<number | undefined>();
  const { app } = useSisenseContext();

  const { styleOptions, dataSource = app?.defaultDataSource, dataOptions } = props;

  const defaultSize = getWidgetDefaultSize('pivot', {
    hasHeader: !styleOptions?.header?.hidden,
  });
  const { width, height, ...styleOptionsWithoutSizing } = props.styleOptions || {};

  if (!dataOptions) {
    return null;
  }

  return (
    <DynamicSizeContainer
      defaultSize={defaultSize}
      size={{
        width: width,
        height: styleOptions?.isAutoHeight ? calcPivotTableWidgetHeight(pivotTableHeight) : height,
      }}
    >
      <WidgetContainer
        {...props}
        dataSetName={dataSource && getDataSourceName(dataSource)}
        onRefresh={() => setRefreshCounter(refreshCounter + 1)}
      >
        <PivotTable
          dataSet={props.dataSource}
          dataOptions={props.dataOptions}
          styleOptions={styleOptionsWithoutSizing}
          filters={props.filters}
          highlights={props.highlights}
          refreshCounter={refreshCounter}
          onHeightChange={setPivotTableHeight}
        />
      </WidgetContainer>
    </DynamicSizeContainer>
  );
});
