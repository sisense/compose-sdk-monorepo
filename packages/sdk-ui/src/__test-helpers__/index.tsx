/**
 * Common test utilities
 */
import { PropsWithChildren, ReactElement } from 'react';
import { render, RenderResult } from '@testing-library/react';
import type { Cell, Data } from '@ethings-os/sdk-data';
import isObject from 'lodash-es/isObject';
import userEvent from '@testing-library/user-event';
import { SisenseContext, SisenseContextPayload } from '../sisense-context/sisense-context';
import { Authenticator, HttpClient } from '@ethings-os/sdk-rest-client';
import { ClientApplication } from '../app/client-application';
import { DeepPartial } from 'ts-essentials';
import { useTranslation } from 'react-i18next';
import { translation } from '../translation/resources/en';
import { CustomWidgetsProvider } from '../custom-widgets-provider';
import { MenuProvider } from '@/common/components/menu/menu-provider';

type UserSetupFn = (typeof userEvent)['setup'];
interface SetupResult extends RenderResult {
  user: ReturnType<UserSetupFn>;
}
/**
 * This helper method combines the userEvent.setup() and render() calls at the
 * start of a unit test using @testing-library/react.
 *
 * Based on recommendation by: https://testing-library.com/docs/user-event/intro/#writing-tests-with-userevent
 *
 * @param jsx - React component instance to test, passed directly to render()
 * @param useFakeTimers - true if Vitest fake timers are enabled
 * @returns
 */
export function setup(jsx: ReactElement, useFakeTimers = false): SetupResult {
  return {
    // Need to set "advanceTimers" if we're using Vitest's fake timers.
    // The solution in https://github.com/testing-library/user-event/issues/833#issuecomment-1171452841
    // is for Jest, but this seems to work for Vitest as well.
    user: userEvent.setup(
      useFakeTimers ? { advanceTimers: (ms) => vi.advanceTimersByTime(ms) } : undefined,
    ),
    ...render(jsx),
  };
}

export const MockedHighchartsReact = ({
  options,
}: {
  /**
   * Highcharts chart configuration object.
   * Please refer to the Highcharts (API documentation)[https://api.highcharts.com/highcharts/].
   */
  options: object;
}) => {
  return <div>{JSON.stringify(options)}</div>;
};

type MockedSisenseContextProviderProps = PropsWithChildren<
  DeepPartial<Omit<SisenseContextPayload, 'app'>>
>;

const mockHttpClient = new HttpClient(
  'http://mock-url/sometenant?someparam=true',
  {} as Authenticator,
  'mock-env',
);
export const MockedSisenseContextProvider = ({
  children,
  tracking = {
    enabled: true,
  },
  isInitialized = true,
  errorBoundary,
}: MockedSisenseContextProviderProps) => {
  return (
    <SisenseContext.Provider
      value={{
        app: {
          httpClient: mockHttpClient,
        } as ClientApplication,
        tracking: {
          enabled: tracking.enabled ?? true,
          packageName: tracking.packageName ?? 'sdk-ui',
        },
        isInitialized,
        errorBoundary: { showErrorBox: true, ...errorBoundary },
      }}
    >
      <MenuProvider>
        <CustomWidgetsProvider>{children}</CustomWidgetsProvider>
      </MenuProvider>
    </SisenseContext.Provider>
  );
};

/**
 * Extends the provided data by blurring specified rows.
 *
 * @param {Data} data - The original data object.
 * @param {number[]} blurRowsIndexes - An array of row indexes to be blurred.
 * @returns {Data} The data object with blurred rows.
 */
export const withBlurredRows = (data: Data, blurRowsIndexes: number[]) => ({
  ...data,
  rows: data.rows.map((row, rowIndex) =>
    row.map((value) => {
      const cell: Cell = isObject(value) ? value : { data: value };

      if (blurRowsIndexes.includes(rowIndex)) {
        return { ...cell, blur: true };
      }
      return cell;
    }),
  ),
});

/**
 * Mocking the i18next translation functionality for tests.
 */
export const setupI18nMock = () => {
  vi.mock('react-i18next', async () => {
    const original: { useTranslation: typeof useTranslation } = await vi.importActual(
      'react-i18next',
    );

    // Helper function to get nested translations safely
    const getNestedTranslation = (key: string): string | undefined => {
      return key.split('.').reduce((accumulator: any, currentKey: string) => {
        if (accumulator && typeof accumulator === 'object' && currentKey in accumulator) {
          return accumulator[currentKey];
        }
        return undefined;
      }, translation) as string | undefined;
    };

    return {
      ...original,
      useTranslation: () => ({
        ...original.useTranslation(),
        t: vi.fn((key, options?: Record<string, any>) => {
          const translatedValue = getNestedTranslation(key);
          if (translatedValue && options) {
            // Replace variables in the translated string
            return translatedValue.replace(/{{(.*?)}}/g, (_, varName) => {
              const trimmedVarName = varName.trim();
              return options[trimmedVarName] || `{{${trimmedVarName}}}`; // Fallback if variable is not found
            });
          }
          return translatedValue ?? key; // Fallback to the key if translation is not found
        }),
      }),
    };
  });
};
