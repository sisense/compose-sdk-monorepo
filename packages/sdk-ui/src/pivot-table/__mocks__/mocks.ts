import * as DM from '@/__test-helpers__/sample-ecommerce';
import { filterFactory, measureFactory } from '@sisense/sdk-data';
import { PivotTableProps } from '../../props';

export const mockPivotTableProps: PivotTableProps = {
  dataSet: DM.DataSource,
  dataOptions: {
    rows: [
      {
        column: DM.Category.Category,
        includeSubTotals: true,
        sortType: 'sortAsc',
      },
      {
        column: DM.Commerce.AgeRange,
        includeSubTotals: true,
        sortType: {
          direction: 'sortDesc',
          by: {
            valuesIndex: 0,
            columnsMembersPath: ['Male'],
          },
        },
      },
      DM.Commerce.Condition,
    ],
    columns: [{ column: DM.Commerce.Gender, includeSubTotals: true }],
    values: [
      {
        column: measureFactory.sum(DM.Commerce.Cost, 'Total Cost'),
        dataBars: true,
        totalsCalculation: 'sum',
      },
      {
        column: measureFactory.sum(DM.Commerce.Revenue, 'Total Revenue'),
        totalsCalculation: 'sum',
      },
    ],
    grandTotals: { title: 'Grand Total', rows: true, columns: true },
  },
  filters: [filterFactory.members(DM.Commerce.Gender, ['Female', 'Male'])],
};
