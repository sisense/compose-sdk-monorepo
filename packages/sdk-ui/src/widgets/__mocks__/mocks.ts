import { QueryResultData } from '@sisense/sdk-data';

import * as DM from '../../__test-helpers__/sample-ecommerce';
import { mockPivotTableProps } from '../../pivot-table/__mocks__/mocks';
import { ChartWidgetProps, PivotTableWidgetProps } from '../../props';

export const mockPivotTableWidgetProps: PivotTableWidgetProps = {
  ...mockPivotTableProps,
  dataSource: DM.DataSource,
  title: 'Pivot table widget title',
  description: 'Pivot table widget desription',
  styleOptions: {
    width: 800,
    membersColor: true,
    headersColor: true,
    border: true,
    borderColor: 'lightgrey',
  },
} as PivotTableWidgetProps;

export const mockChartWidgetPropsForTable: ChartWidgetProps = {
  chartType: 'table',
  dataSource: DM.DataSource,
  title: 'Chart widget title',
  description: 'Chart widget desription',
  dataOptions: {
    columns: [DM.Commerce.AgeRange, DM.Commerce.Revenue],
  },
};

export const mockResolvedQuery: QueryResultData = {
  columns: [
    {
      name: 'Age Range',
      type: 'text',
    },
    {
      name: 'Revenue',
      type: 'number',
    },
  ],
  rows: [
    [
      {
        data: '19-24',
        text: '19-24',
        blur: false,
      },
      {
        data: 43.98411178588867,
        text: '43.9841117858887',
        blur: false,
      },
    ],
    [
      {
        data: '65+',
        text: '65+',
        blur: false,
      },
      {
        data: 62.803104400634766,
        text: '62.8031044006348',
        blur: false,
      },
    ],
  ],
};
