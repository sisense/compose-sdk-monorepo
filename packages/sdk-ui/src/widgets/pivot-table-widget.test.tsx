/** @vitest-environment jsdom */
import { EMPTY_PIVOT_QUERY_RESULT_DATA } from '@sisense/sdk-data';
import { render } from '@testing-library/react';

import { setupI18nMock } from '@/__test-helpers__';
import { createMockPivotClient } from '@/pivot-table/__mocks__/pivot-client-mock';
import { SisenseContextPayload } from '@/sisense-context/sisense-context';

import { ClientApplication } from '../app/client-application';
import { executePivotQueryMock } from '../query/__mocks__/execute-query';
import { useSisenseContextMock } from '../sisense-context/__mocks__/sisense-context';
import { mockPivotTableWidgetProps } from './__mocks__/mocks';
import { PivotTableWidget } from './pivot-table-widget';

setupI18nMock();

vi.mock('../query/execute-query');
vi.mock('../sisense-context/sisense-context');

describe('PivotTableWidget', () => {
  beforeEach(() => {
    const mockPivotClient = createMockPivotClient();

    const contextMock: SisenseContextPayload = {
      app: {
        pivotClient: mockPivotClient,
        settings: {
          trackingConfig: {
            enabled: false,
          },
        },
      } as ClientApplication,
      isInitialized: true,
      tracking: {
        enabled: false,
        packageName: '',
      },
      errorBoundary: {
        showErrorBox: true,
      },
    };
    useSisenseContextMock.mockReturnValue(contextMock);
  });

  it('should render empty pivot table widget', async () => {
    executePivotQueryMock.mockResolvedValue(EMPTY_PIVOT_QUERY_RESULT_DATA);
    const { container, findByLabelText } = render(
      <PivotTableWidget {...mockPivotTableWidgetProps} />,
    );
    const pivotTable = await findByLabelText('pivot-table-root');
    expect(pivotTable).toBeTruthy();
    expect(container).toMatchSnapshot();
  });
});
