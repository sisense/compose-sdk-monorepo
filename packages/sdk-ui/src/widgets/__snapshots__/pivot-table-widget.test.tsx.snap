// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PivotTableWidget > should render empty pivot table widget 1`] = `
<div>
  <div
    style="width: 800px; height: 100%;"
  >
    <div
      style="width: 800px; height: 525px; position: relative;"
    >
      <div
        style="position: absolute; top: 0px; right: 0px; bottom: 0px; left: 0px;"
      >
        <div
          class="csdk-w-full csdk-h-full csdk-overflow-hidden"
        >
          <div
            class="csdk-h-full"
            style="padding: 0px;"
          >
            <div
              class="csdk-h-full csdk-overflow-hidden"
              style="border-width: 1px; border-color: lightgrey; border-radius: 0; box-shadow: none; display: flex; flex-direction: column;"
            >
              <div
                class="csdk-flex csdk-flex-row csdk-items-center csdk-px-2 csdk-min-h-[32px]"
                style="background-color: rgb(255, 255, 255);"
              >
                <div
                  class="csdk-text-left csdk-w-full csdk-whitespace-nowrap csdk-overflow-hidden"
                  style="color: rgb(0, 0, 0); font-family: "Open Sans","Roboto","Helvetica","Arial",sans-serif; font-size: 15px;"
                >
                  Pivot table widget title
                </div>
                <div
                  class="csdk-ml-auto csdk-flex csdk-items-center"
                >
                  <button
                    aria-label="Click to view full details"
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-av538e-MuiButtonBase-root-MuiIconButton-root"
                    data-mui-internal-clone-element="true"
                    style="background-color: rgb(255, 255, 255);"
                    tabindex="0"
                    type="button"
                  >
                    <svg
                      height="24"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M11.5 20a7.5 7.5 0 1 1 0-15 7.5 7.5 0 0 1 0 15zm0-1a6.5 6.5 0 1 0 0-13 6.5 6.5 0 0 0 0 13zm-.5-8h1v5h-1v-5zm0-1v-.998h1V10h-1z"
                        fill="#000000"
                      />
                    </svg>
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </div>
              </div>
              <div
                style="flex-grow: 1; min-width: 0; min-height: 0; background-color: rgb(255, 255, 255);"
              >
                <div
                  style="width: 100%; height: 100%;"
                >
                  <div
                    style="width: 400px; height: 525px; position: relative;"
                  >
                    <div
                      style="position: absolute; top: 0px; right: 0px; bottom: 0px; left: 0px;"
                    >
                      <div
                        class="csdk-relative csdk-h-full"
                        id="overlay"
                      >
                        <div
                          aria-label="pivot-table-root"
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
